hugr-passes/src/dataflow/total_context.rs

```diff
@@ -0,0 +1,54 @@
+use std::hash::Hash;
+
+use ascent::lattice::BoundedLattice;
+use hugr_core::{ops::OpTrait, Hugr, HugrView, IncomingPort, Node, OutgoingPort, PortIndex};
+
+use super::partial_value::{PartialValue, Sum};
+use super::{BaseValue, DFContext};
+
+/// A simpler interface like [DFContext] but where the context only cares about
+/// values that are completely known (as `V`s), i.e. not `Bottom`, `Top`, or
+/// Sums of potentially multiple variants.
+pub trait TotalContext<V>: Clone + Eq + Hash + std::ops::Deref<Target = Hugr> {
```

Comment:
This may belong in the follow-up PR (even tho it's fairly general, so I think lives in this module, it's only used there)...